pkg/planner/core/plan_cost_ver2.go

```diff
@@ -170,31 +170,48 @@ func (p *PhysicalTableScan) GetPlanCostVer2(taskType property.TaskType, option *
 	}
 	hasFullRangeScan := ranger.HasFullRange(p.Ranges, unsignedIntHandle)
 
-	// Apply TiFlash startup cost to prefer TiKV for small table scans
 	if p.StoreType == kv.TiFlash {
+		// Apply TiFlash startup cost to prefer TiKV for small table scans
 		p.PlanCostVer2 = costusage.SumCostVer2(p.PlanCostVer2, scanCostVer2(option, TiFlashStartupRowPenalty, rowSize, scanFactor))
-	} else if !p.isChildOfIndexLookUp && hasFullRangeScan {
-		newRowCount := getTableScanPenalty(p, rows)
-		if newRowCount > 0 {
-			p.PlanCostVer2 = costusage.SumCostVer2(p.PlanCostVer2, scanCostVer2(option, newRowCount, rowSize, scanFactor))
+		// Add penalty for late materialization
+		if len(p.LateMaterializationFilterCondition) > 0 {
+			// For late materialization, we will scan all rows of filter columns in the table or given range first.
+			// And the rows of rest columns after filtering may be discrete, so we need to add some penalty for it.
+			// so late materialization cost = lmRowSize * scanFactor * scanRowCount + restRowSize * scanFactor * rowCount * lateMaterializationFactor
+			cols := make([]*expression.Column, 0, len(p.Columns))
+			cols = expression.ExtractColumnsFromExpressions(cols, p.LateMaterializationFilterCondition, nil)
+			lmRowSize := getAvgRowSize(p.StatsInfo(), cols)
+			p.PlanCostVer2 = costusage.MulCostVer2(p.PlanCostVer2, lmRowSize/rowSize*defaultVer2Factors.LateMaterializationScan.Value)
```

Comment:
It is wrong. So I cannot see any `LateMaterializationScan` factor in the cost tracer.